test-it/karma/javascriptinvoke/Async.Test.js

```diff
@@ -297,41 +297,37 @@ describe('A JavaScript function invoke', function () {
                 const {rawPrint, rawPrintError} = await runSlicesAsync();
                 expect(rawPrint).toBeEmptyString();
                 expect(rawPrintError).toBeEmptyString();
-                expect(viPathParser('error.status')).toBeFalse();
-                expect(viPathParser('error.code')).toBe(0);
-                expect(viPathParser('error.source')).toBeEmptyString();
-                expect(viPathParser('return')).toBe(49);
-                expect(error.message).toMatch(/retrieved more than once/);
+                expect(viPathParser('error.status')).toBeTrue();
+                expect(viPathParser('error.code')).toBe(44300);
+                expect(viPathParser('error.source')).toMatch(/retrieved more than once/);
+                expect(viPathParser('return')).toBe(0);
+                expect(error.message).toMatch(/threw an error, so this callback cannot be invoked/);
             };
         });
         it('using the completion callback', async function () {
             window.SingleFunction = function (input, jsapi) {
                 var completionCallback = jsapi.getCompletionCallback();
-                try {
-                    jsapi.getCompletionCallback();
-                } catch (ex) {
-                    error = ex;
-                }
+                jsapi.getCompletionCallback();
```

Comment:
Does removing the try catch mean this no longer results in an exception?

I think it still should be resulting in an exception and also be reporting the error as a LabVIEW error. That way the user callstack doesn't keep running after an exception occurs.

I think it ends up looking like this:
1. enter lv function
2. enter user js function
3. first call to api good
4. second call to api results in error (called completioncb twice), lv error merged, lvfunctionstatefinished, lv node occurence, throws error on callstack so user code observes exception
5. If the user doesn't catch exceptions (which they really shouldn't) then our wrapper catches the exception, but lvfunctionstatefinished now set, So, if internal rethrow and break vireo and the callstack, but if not internal then console.error the message. We don't want to have an exception like that to break the vireo callstack.

That might be introducing another piece of state that is captured by reportErrorFromRuntime / reportErrorFromRuntimeAsync and reportErrorFromUser / reportErrorFromUserAsync.

reportingErrorFromRuntime throws for internal but not for noninternal, it just console.errors
reportingErrorFromUser throws for both internal and not internal to break the users callstack when they do something resulting in a jsapi error.

Reply:
You want it to throw an exception **and** put an error on the wire? Won't that be confusing to users, since they probably expect error handling to prevent the console logging of that type of error?

Reply:
Yea I think it should not be the jsapi call that is doing the console.error, that should just throw the exception.

It is our code that is invoking the user's function that should take all uncaught exceptions, console.error them, and put them on the LabVIEW error wire, and swallow them to prevent breaking our JS callstack (unless it is internal, etc etc).

When I was working on Christina's demo it was really frustrating to look at the labview error dialog for JS errors because right now we swallow them and just put them on the labview error wire. The benefit of console.error messages is that it will include the callstack and clickable line numbers to jump to the error location is source, etc.

Reply:
Let me make sure I get this:

In the jsapi calls (getCompletionCallback and completionCallback, but also reportExecutionError and updateReturnValue?), we should throw exceptions after setting the occurrence so that execution doesn't get stopped. It should not write to the error wire.

We should catch these exceptions in the actual jsJavaScriptInvoke call. Then we put them on the error wire and console.error them as well, but don't rethrow any unless it's an internal function call.

Is that correct?

For the calls to reportExecutionError and updateReturnValue after the try/catch in jsJavaScriptInvoke, should we add a try/catch to do the same?